py/persistentcode.c

```diff
@@ -164,51 +178,167 @@ STATIC void load_bytecode_qstrs(mp_reader_t *reader, byte *ip, byte *ip_top) {
     }
 }
 
+#if MICROPY_EMIT_X64
+STATIC void link_qstr(uint8_t *pc, bool is_obj, qstr qst) {
+    uint64_t val = qst;
+    if (is_obj) {
+        val = (mp_uint_t)MP_OBJ_NEW_QSTR(qst);
+    }
+    pc[0] = val & 0xff;
+    pc[1] = (val >> 8) & 0xff;
+    pc[2] = (val >> 16) & 0xff;
+    pc[3] = (val >> 24) & 0xff;
+}
+#elif MICROPY_EMIT_X86
+STATIC void link_qstr(uint8_t *pc, bool is_obj, qstr qst) {
+    uint32_t val = qst;
+    if (is_obj) {
+        val = (mp_uint_t)MP_OBJ_NEW_QSTR(qst);
+    }
+    pc[0] = val & 0xff;
+    pc[1] = (val >> 8) & 0xff;
+    pc[2] = (val >> 16) & 0xff;
+    pc[3] = (val >> 24) & 0xff;
+}
+#elif MICROPY_EMIT_THUMB
+STATIC void asm_thumb_rewrite_mov(uint8_t *pc, uint16_t val) {
+    // high part
+    *(uint16_t*)pc = (*(uint16_t*)pc & 0xfbf0) | (val >> 1 & 0x0400) | (val >> 12);
+    // low part
+    *(uint16_t*)(pc + 2) = (*(uint16_t*)(pc + 2) & 0x0f00) | (val << 4 & 0x7000) | (val & 0x00ff);
+
+}
+STATIC void link_qstr(uint8_t *pc, bool is_obj, qstr qst) {
+    if (is_obj) {
+        // qstr object, movw and movt
+        mp_uint_t o = (mp_uint_t)MP_OBJ_NEW_QSTR(qst);
+        asm_thumb_rewrite_mov(pc, o); // movw
+        asm_thumb_rewrite_mov(pc + 4, o >> 16); // movt
+    } else {
+        // qstr number, movw instruction
+        asm_thumb_rewrite_mov(pc, qst); // movw
+    }
+}
+#elif MICROPY_EMIT_XTENSA
+STATIC void link_qstr(uint8_t *pc, bool is_obj, qstr qst) {
+    uint32_t val = qst;
+    if (is_obj) {
+        val = (mp_uint_t)MP_OBJ_NEW_QSTR(qst);
+    }
+    *(uint32_t*)pc = val;
+}
+#endif
+
 STATIC mp_raw_code_t *load_raw_code(mp_reader_t *reader) {
-    // load bytecode
-    size_t bc_len = read_uint(reader);
-    byte *bytecode = m_new(byte, bc_len);
-    read_bytes(reader, bytecode, bc_len);
+    // load function kind
+    int kind = read_byte(reader);
```

Comment:
Perhaps it's possible to combine "kind" with some other flags into this byte?

Reply:
Ok, since it's only 2 bits I've now combined it with the number of bytes in the bytecode/function data, like: `write_uint((len << 2) | (kind - 2))`